bjoern/server.c

```diff
@@ -123,14 +141,51 @@ ev_signal_on_sigint(struct ev_loop* mainloop, ev_signal* watcher, const int even
   ev_cleanup_init(cleanup_watcher, pyerr_set_interrupt);
   ev_cleanup_start(mainloop, cleanup_watcher);
 
-  ev_io_stop(mainloop, &((ThreadInfo*)ev_userdata(mainloop))->accept_watcher);
+  ThreadInfo *thread_info = (ThreadInfo*)ev_userdata(mainloop);
+
+  ev_io_stop(mainloop, &thread_info->accept_watcher);
   ev_signal_stop(mainloop, watcher);
 #ifdef WANT_SIGNAL_HANDLING
   ev_timer_stop(mainloop, &timeout_watcher);
 #endif
+#ifdef WANT_GRACEFUL_SHUTDOWN
+  ev_signal_stop(mainloop, &thread_info->sigterm_watcher);
+  ev_timer_stop(mainloop, &timeout_watcher);
```

Comment:
Good point! I'll move the timeout watcher up to the top and shuffle the code around so that it only gets stopped once.

From the documentation for `ev_TYPE_stop` here http://pod.tst.eu/http://cvs.schmorp.de/libev/ev.pod, it says "Stops the given watcher if active, and clears the pending status (whether the watcher was active or not)." so it looks like it's safe to stop it more than once. Because it also clears the pending status means that we shouldn't receive a `timeout_watcher` callback during this handler either.